apollo-federation/src/composition/satisfiability/validation_state.rs

```diff
@@ -221,35 +248,339 @@ impl ValidationState {
+    ) -> bool {
+        let previous_visits_for_vertex = previous_visits.entry(supergraph_path_tail).or_default();
+
+        for previous_visit in previous_visits_for_vertex.iter() {
+            // Check if current visit is a superset of the previous visit (isSupersetOrEqual in JS)
+            if current_vertex_visit
+                .subgraph_context_keys
+                .is_superset(&previous_visit.subgraph_context_keys)
+                && previous_visit
+                    .override_conditions
+                    .iter()
+                    .all(|(label, is_enabled)| {
+                        current_vertex_visit.override_conditions.get(label) == Some(is_enabled)
+                    })
+            {
+                // This means that we've already seen the type we're currently on in the supergraph,
+                // and when saw it we could be in one of `previousSources`, and we validated that
+                // we could reach anything from there. We're now on the same type, and have strictly
+                // more options regarding subgraphs. So whatever comes next, we can handle in the exact
+                // same way we did previously, and there is thus no way to bother.
+                return true;
+            }
+        }
+
+        // We're gonna have to validate, but we can save the new set of sources here to hopefully save work later.
+        previous_visits_for_vertex.push(current_vertex_visit.clone());
+        false
+    }
+
+    /// Returns whether the entire entire visit to this state can be skipped. If the state is
+    /// partitioned, note that each individual partition must be skippable for the state to be
+    /// skippable.
+    ///
+    /// This corresponds to `canSkipVisit` in the JavaScript implementation.
+    pub(super) fn can_skip_visit(
+        &self,
+        previous_visits: &mut IndexMap<petgraph::graph::NodeIndex, Vec<NodeVisit>>,
+    ) -> Result<bool, FederationError> {
+        let vertex = self.supergraph_path.tail();
+
+        match &self.subgraph_path_infos {
+            SubgraphPathInfos::Paths(paths) => {
+                // Non-partitioned case
+                let current_vertex_visit = NodeVisit {
+                    subgraph_context_keys: self.current_subgraph_context_keys()?,
+                    override_conditions: self.selected_override_conditions.clone(),
+                };
+                Ok(Self::can_skip_visit_for_subgraph_paths(
+                    vertex,
+                    paths,
+                    &current_vertex_visit,
+                    previous_visits,
+                ))
+            }
+            SubgraphPathInfos::TopLevelMutationField { paths, .. } => {
+                // Partitioned case
+                let mut can_skip = true;
+                for subgraph_path_infos in paths.values() {
+                    let current_vertex_visit = NodeVisit {
+                        subgraph_context_keys: self.current_subgraph_context_keys()?,
```

Comment:
Here, `current_subgraph_context_keys` takes a different `SubgraphPathInfos` from `self.subgraph_path_infos`. That's why `current_subgraph_context_keys` was changed to take an argument.